2nd assignment

katabana opened the pull request.

katabana committed:
first assignment

katabana committed:
tolower added

katabana committed:
mistakes repaired

katabana committed:
assignment2 Q1

katabana committed:
q1 comment

katabana committed:
q2 done

Brusi commented on assignment2/q1.cpp:
```diff
@@ -0,0 +1,104 @@
+    if(root->value == key)
+        return root;
+
+    Node *l = NULL;
```
No need to declare these early. You can declare them both inside the scope you use them:
```
if(root->left != NULL) {
  Node* l = FindKey(root->left, key);
  if(l != NULL)
    return l;
}
```

Then you can even give them the same name, e.g. `target` or something more indicative than `l`/`r`. Your choice :)

Brusi commented on assignment2/q1.cpp:
```diff
@@ -0,0 +1,104 @@
+    if(descendant != NULL) {
+        cout << "Descendant key is "<< descendant->value <<endl;
+        if(descendant->parent == NULL)
+            cout<< "NULL";
```
In case there are no ancestors, it is ok, in my opinion, to print an 'empty list', i.e. print nothing.
Can skip this print in my opinion (the `while` already handles the no-ancestors case).

Brusi commented on assignment2/q2.cpp:
```diff
@@ -0,0 +1,148 @@
+    Node *left, *right;
+};
+
+bool FindNode(Node *root, Node *to_find) {
```
The name `FindNode` (similarly to FindKey) indicates that you look for something in the tree and return it.
A better name for something that returns `bool` is `NodeExists` or `DoesNodeExistInTree`.
Or even `IsAncestorOf` (because `root` is ancestor of `to_find`).

Brusi commented on assignment2/q2.cpp:
```diff
@@ -0,0 +1,148 @@
+        return FindCommonAncestor(root->right, first, second);
+
+    //if nodes are one in left, other in right tree
+    else
```
No need for this `else`, since there is a `return` in the `if` before it.
You did rely on the early returns in the previous cases of this function, no reason not to rely on it here too.

Brusi commented on assignment2/q2.cpp:
```diff
@@ -0,0 +1,148 @@
+        return NULL;
+
+    //nodes are not in the tree
+    if(!FindNode(root, first) || !FindNode(root, second))
```
Idea for optimization (not mandatory):
You can simulate the solution of `FindNode` in `root` by using the results from `FindNode` in the left and right children, and checking the root node itself.
This will save you searching the tree 4 times each call (instead of only 2, once for each first/second node)

Brusi commented on assignment2/q2.cpp:
```diff
@@ -0,0 +1,148 @@
+
+//O(N^2) without any additional memory or changing structure
+//could be O(N) if i.e. it stores paths to nodes and then compares them
+Node *FindCommonAncestor(Node *root, Node *first, Node *second) {
```
Nice solution, straightforward and readable. And well commented too! :+1: 
As you wrote, you could improve the complexity in various ways.

Brusi reviewed (approved):
Well done!

You can add the rest of the team as collaborators so they can peer-review your code. Ther github names are:
elenial
matinehAkhlaghinia
maryb14
TashaSekularac
Mais051

katabana committed:
not tested digitally

I would if I had more time to write my implementations of isWord and isPrefix
However I did test it in thoughts
